lib/active_record_view_model/update_data.rb

```diff
@@ -285,6 +287,19 @@ def parse(hash_data, valid_reference_keys)
                   UpdateData.verify_schema!(Schemas::COLLECTION_UPDATE, value)
                   functional_updates = value[ACTIONS_ATTRIBUTE].map do |action|
                     type   = FunctionalUpdate::Type.parse!(action[TYPE_ATTRIBUTE])
+                    values = action[VALUES_ATTRIBUTE]
+
+                    # There's no reasonable interpretation of a remove update that includes data.
+                    # Report it as soon as we detect it.
+                    if type == FunctionalUpdate::Type::Remove
+                      invalid_entries = values.reject { |h| reference_only_hash?(h) }
```

Comment:
polymorphic: as you point out, that's not a supported has_many type
has_many_through: entirely different codepath, will make a seperate pull request for this after merging this.